File: backend-nest-backup/src/notifications/notifications.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
  Query,
  UseGuards,
} from '@nestjs/common';

import { AdminAuthGuard } from '../admin-auth/admin-auth.guard';
import { Roles } from '../admin-auth/roles.decorator';
import { RolesGuard } from '../admin-auth/roles.guard';
import { CreateNotificationDto, UpdateNotificationDto } from './notifications.types';
import { NotificationsService } from './notifications.service';

@Controller('notifications')
export class NotificationsController {
  constructor(private readonly notificationsService: NotificationsService) {}

  @Get()
  list(
    @Query('active') active?: string,
    @Query('city') city?: string,
    @Query('region') region?: string,
  ) {
    const activeOnly = active === 'true' || active === '1';
    return this.notificationsService.list({ activeOnly, city, region });
  }

  @Post()
  @UseGuards(AdminAuthGuard, RolesGuard)
  @Roles('superadmin', 'admin')
  create(@Body() dto: CreateNotificationDto) {
    return this.notificationsService.create(dto);
  }

  @Patch(':id')
  @UseGuards(AdminAuthGuard, RolesGuard)
  @Roles('superadmin', 'admin')
  update(@Param('id') id: string, @Body() dto: UpdateNotificationDto) {
    return this.notificationsService.update(id, dto);
  }

  @Delete(':id')
  @UseGuards(AdminAuthGuard, RolesGuard)
  @Roles('superadmin', 'admin')
  remove(@Param('id') id: string) {
    return this.notificationsService.remove(id);
  }
}

File: backend/src/main/java/com/unbora/api/domain/notification/NotificationsController.java
package com.unbora.api.domain.notification;

import com.unbora.api.domain.notification.dto.CreateNotificationDto;
import com.unbora.api.domain.notification.dto.NotificationRecordDto;
import com.unbora.api.domain.notification.dto.UpdateNotificationDto;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/notifications")
@Tag(name = "Notifications", description = "Notificações e avisos da plataforma")
public class NotificationsController {

    private final NotificationsService notificationsService;

    public NotificationsController(NotificationsService notificationsService) {
        this.notificationsService = notificationsService;
    }

    @GetMapping
    @Operation(summary = "Listar notificações com filtros")
    public List<NotificationRecordDto> list(
            @RequestParam(required = false) String active,
            @RequestParam(required = false) String city,
            @RequestParam(required = false) String region
    ) {
        boolean activeOnly = "true".equalsIgnoreCase(active) || "1".equals(active);
        return notificationsService.list(activeOnly, city, region);
    }

    @PostMapping
    @PreAuthorize("hasAnyRole('SUPERADMIN', 'ADMIN')")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Criar notificação (Admin)")
    public NotificationRecordDto create(@Valid @RequestBody CreateNotificationDto dto) {
        return notificationsService.create(dto);
    }

    @PatchMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPERADMIN', 'ADMIN')")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Atualizar notificação (Admin)")
    public NotificationRecordDto update(@PathVariable String id, @RequestBody UpdateNotificationDto dto) {
        return notificationsService.update(id, dto);
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPERADMIN', 'ADMIN')")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Remover notificação (Admin)")
    public Map<String, Object> remove(@PathVariable String id) {
        return notificationsService.remove(id);
    }
}
